sdlib/d/gc/arena.d

```diff
@@ -751,3 +814,104 @@ unittest resizeLarge {
 	// Now full again:
 	assert(pd9.extent.hpd.full);
 }
+
+unittest resizeLargeGrow {
+	import d.gc.arena;
+	shared Arena arena;
+
+	auto base = &arena.base;
+	scope(exit) arena.base.clear();
+
+	import d.gc.emap;
+	static shared ExtentMap emap;
+	emap.tree.base = base;
+
+	import d.gc.region;
+	shared RegionAllocator regionAllocator;
+	regionAllocator.base = base;
+
+	arena.regionAllocator = &regionAllocator;
+
+	PageDescriptor makeAllocLarge(uint pages) {
+		auto ptr = arena.allocLarge(&emap, pages * PageSize);
+		assert(ptr !is null);
+		auto pd = emap.lookup(ptr);
+		assert(pd.extent.address is ptr);
+		assert(pd.extent.size == pages * PageSize);
+		return pd;
+	}
+
+	void checkGrowLarge(PageDescriptor pd, uint pages) {
+		auto e = pd.extent;
+
+		assert(arena.resizeLarge(&emap, e, pages * PageSize));
+		assert(e.size == pages * PageSize);
+
+		// Confirm that the page after the end of the extent is not included in the map:
+		auto pdAfter = emap.lookup(e.address + e.size);
+		assert(pdAfter.extent !is e);
+
+		// Confirm that the extent correctly grew and remapped:
+		for (auto p = e.address; p < e.address + e.size; p += PageSize) {
+			auto probe = emap.lookup(p);
+			assert(probe.extent == e);
+			assert(probe.data == pd.data);
+			pd = pd.next();
+		}
+	}
+
+	// Allocation 0: 35 pages:
+	auto pd0 = makeAllocLarge(35);
+
+	// Allocation 1: 64 pages:
+	auto pd1 = makeAllocLarge(64);
+	assert(pd1.extent.address == pd0.extent.address + pd0.extent.size);
+
+	// Allocation 2: 128 pages:
+	auto pd2 = makeAllocLarge(128);
+	assert(pd2.extent.address == pd1.extent.address + pd1.extent.size);
+
+	// Try to grow allocation 0, but cannot, there is no space to grow:
+	assert(!arena.resizeLarge(&emap, pd0.extent, 36 * PageSize));
+
+	// Try to grow allocation 0 by zero; this will succeed:
+	assert(arena.resizeLarge(&emap, pd0.extent, 35 * PageSize));
+
+	// Try to grow allocation 2, but cannot, asked for one too many pages:
+	assert(!arena.resizeLarge(&emap, pd2.extent, 414 * PageSize));
+
+	// Grow allocation 2 successfully by 413 pages:
+	checkGrowLarge(pd2, 413);
+
+	// Free allocation 1:
+	arena.free(&emap, pd1, pd1.extent.address);
+
+	// Grow allocation 0:
+	checkGrowLarge(pd0, 44);
+
+	// Try to grow allocation 0 but fail:
+	assert(!arena.resizeLarge(&emap, pd0.extent, uint.max * PageSize));
+	assert(!arena.resizeLarge(&emap, pd0.extent, 9999 * PageSize));
+	assert(!arena.resizeLarge(&emap, pd0.extent, 100 * PageSize));
+
+	// Grow allocation 0 to largest size for which there is room :
+	checkGrowLarge(pd0, 99);
+
+	// Confirm that hpd is full:
+	assert(pd0.extent.hpd.full);
```

Comment:
These comment on each line really are redundant and prevent the code from being grouped in chunks that make sense together.

Think about reading a book without any paragraphs in it, just sentences on their own line.